### PR TITLE
perf: performance improvements in respect of large baskets

### PR DESCRIPTION
**Pull request by SGrueber**
## PR Type

[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no API changes)
[ ] Build-related changes
[ ] CI-related changes
[ ] Documentation content changes
[ ] Application / infrastructure changes
[ ] Other: 

## What Is the Current Behavior?
In case of large basket ( > 100 items ) the response times after user interactions are unacceptable long, e.g. it takes a really long time to show the mini basket or the screen is blocked to long after adding, updating or deleting basket items.

Issue Number: Closes #

## What Is the New Behavior?
There are optimizations concerning the IAD as well as the technical handling in respect of large baskets.
- display only a limited number of line items in the mini basket
- display a paging bar in case a basket has more than 25 items ( configurable in the line-item-list.component)
- display the loading animation only over the line item list instead of the whole basket page
- addItemToBasket, updateBasketItem, deleteBasketItem action synchronize the basket item in the store immediately after the Success action has been triggered; the subsequent loading of the basket is still necessary to update the basket total
- the debounce of 1s after addItemToBasket has been reduced to 500ms

## Does this PR Introduce a Breaking Change?

[x] Yes
[ ] No

## Other Information

[AB#78155](https://dev.azure.com/intershop-com/cefd1005-00a7-4c79-927f-a16947d1b2e6/_workitems/edit/78155)

### CODE REVIEW

File: src/app/core/store/customer/basket/basket-items.effects.ts [hunk 1]
```diff
@@ -103,6 +109,29 @@ export class BasketItemsEffects {
     )
   );
 
+  /**
+   * Update basket item effect.
```
Comment (MaxKless) [inline src/app/core/store/customer/basket/basket-items.effects.ts:113]: I'm not sure if this comment actually helps anyone :D

File: src/app/core/store/customer/basket/basket-items.effects.ts [hunk 2]
```diff
@@ -103,6 +109,29 @@ export class BasketItemsEffects {
...
+   */
+  updateBasketItem$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(updateBasketItem),
+      mapToPayloadProperty('lineItemUpdate'),
+      withLatestFrom(this.store.pipe(select(getCurrentBasket))),
+      filter(([payload, basket]) => !!basket.lineItems && !!payload),
+      concatMap(([li]) =>
```
Comment (MaxKless) [inline src/app/core/store/customer/basket/basket-items.effects.ts:121]: For clarity's sake, I think renaming this variable to be longer would help.

File: src/app/core/store/customer/basket/basket.reducer.ts [hunk 3]
```diff
@@ -193,13 +200,31 @@ export const basketReducer = createReducer(
...
+  })),
+  on(deleteBasketItemSuccess, (state, action) => ({
+    ...state,
+    basket: { ...state.basket, lineItems: state.basket.lineItems.filter(item => item.id !== action.payload.itemId) },
     info: action.payload.info,
     validationResults: initialValidationResults,
   })),
   on(addItemsToBasketSuccess, (state, action) => ({
     ...state,
+    basket: { ...state.basket, lineItems: unionBy(action.payload.lineItems, state.basket.lineItems || [], 'id') },
```
Comment (MaxKless) [inline src/app/core/store/customer/basket/basket.reducer.ts:227]: It's not going to make a difference here, but I think we should use `||` and `??` in a way that reflects our intentions:
- `??` when you want to catch an undefined or null value
- `||` when you want to use the second value if the first is falsy.

**Comment by schadii** [comment]
Nice work! Tnx Max

**Review by schadii — APPROVED** [review]
Nice work! Tnx Max

**Review by iwiederhold — COMMENTED** [review]
Yes!!!

**Review by iwiederhold — APPROVED** [review]

**Review by MaxKless — APPROVED** [review]

**Pull request merged**